public_dropin_gpu_environments/nemollm_inference_ms/artifact_downloader/datarobot_custom_code/model_artifact_downloader.py

```diff
@@ -0,0 +1,542 @@
+import datetime
+import time
+import os
+import pprint
+import json
+import subprocess
+from functools import partial
+from multiprocessing import Lock, Manager, Pool, Process
+
+from datarobot.mlops.connected.client import MLOpsClient
+from datarobot.mlops.mlops import MLOps
+import datarobot as dr
+
+from datarobot_custom_code.utils import bytes_to_mb_str, calculate_rate, calculate_rate_str, parse_s3_uri, list_zip_contents, \
+    extract_zip_content, load_yaml_from_file, sum_file_sizes, get_disk_space, merge_lists
+
+from datarobot_custom_code.s3_file_download_helper import S3FileDownloadHelper
+
+
+class SupportedStoreType:
+    S3 = "s3"
+
+
+class StoreURIPrefix:
+    S3 = "s3:"
+
+
+class ModelDownloader:
+    REMOTE_FILE_SUFFIX = ".remote"
+    METADATA_FILE = "model-metadata.yaml"
+    MLOPS_RUNTIME_PARAM_PREFIX = "MLOPS_RUNTIME_PARAM_"
+
+    def __init__(self,
+                 deployment_id,
+                 code_dir,
+                 verify_ssl=True,
+                 mb_to_update=50,
+                 seconds_to_update=10,
+                 chunk_size=1024*1024,
+                 nr_processes=1,
+                 verify_checksum=True):
+        """
+        ModelDownloader constructor
+        :param deployment_id:  The deployment to download the model code from
+        :param code_dir: The local code directory to use
+        :param verify_ssl: If True verify SSL certificate
+        """
+        self._deployment_id = deployment_id
+        self._dr_client = dr.Client()
+        self._mlops_service_url = os.getenv("DATAROBOT_ENDPOINT")
+        self._mlops_client = MLOpsClient(service_url=os.getenv("DATAROBOT_ENDPOINT").replace('/api/v2', ''),
+                                         api_key=os.getenv("DATAROBOT_API_TOKEN"),
+                                         verify=verify_ssl)
+        self._mlops = MLOps().set_deployment_id(self._deployment_id).set_api_spooler().init()
+
+        self._code_dir = code_dir
+        self._deployment_info = self._mlops_client.get_deployment(deployment_id)
```

Comment:
Can this code work when there is no deployment? In Custom Model Testing there is no deployment but we still need to download model artifacts (and it's ok if we don't get progress updates in that case 🤷‍♂️ )